### PR TITLE
Multi AZ ASG

### PR DESCRIPTION
**Pull request by tom-butler**
# Why this change is needed

Currently we create a single ASG per AZ and per nodetype, when using both spot and ondemand node types this means you will have 6 ASGs, this choice was made as ASG AZ rebalancing can mess with cluster autoscaler. 

This approach works, but raises additional issues:
- min number of nodes is min number of ASGs
- autoscaling from 0 can be a bit broken, cluster autoscaler will sometimes scale up one asg and leave other ones

# How this change works

To simplify deployments for new instances we now support new x_nodes variables:
- min_nodes
- max_nodes
- desired_nodes
- min_spot_nodes
- max_spot_nodes

The _nodes_per_az variables should still work if _nodes variable is >0 we use that, otherwise we will combine _nodes_per_az by the number of deployed AZs

# Negative effects of this change

This change has been tested without adding the new *_nodes variables, it will modify the values of your nodes-0 and spot-0 to be equivelant to the value of min_nodes_per_az * number of azs, it will remove the exra autoscaling groups, After this change has been made you should probably run the new roll_instances.sh script, to remove the old nodes in your asg which are deployed with an old version of the launch template that is pinned to a single AZ.

### CODE REVIEW

**Comment by tom-butler** [comment]
This should solve #88

File: nodes/modules/workers/workers.tf [hunk 1]
```diff
@@ -1,10 +1,10 @@
 resource "aws_autoscaling_group" "nodes" {
-  count            = var.nodes_enabled ? length(var.nodes_subnet_group) : 0
+  count            = var.nodes_enabled ? 1 : 0
   desired_capacity = var.desired_nodes
   max_size         = var.max_nodes
   min_size         = var.min_nodes
-  name             = "${var.node_group_name}-${aws_launch_template.node[count.index].id}-nodes-${count.index}"
-  vpc_zone_identifier = [element(var.nodes_subnet_group, count.index)]
+  name             = "${var.node_group_name}-${aws_launch_template.node[0].id}-nodes"
+  vpc_zone_identifier = var.nodes_subnet_group
```
Comment (NikitaGandhi) [inline nodes/modules/workers/workers.tf]: `nodes_subnet_group` is a list variable so not sure if this works or requires element index -
`var.nodes_subnet_group[0]`

File: docs/variables.md [hunk 2]
```diff
@@ -32,6 +32,11 @@ This page gives an overview of all possible variables that can be put in a `terr
 | [default_worker_instance_type](#default_worker_instance_type)                               | Nodes                | No  | "" |
 | [group_enabled](#group_enabled)                                                             | Nodes                | No  | false |
 | [spot_nodes_enabled](#spot_nodes_enabled)                                                   | Nodes                | No  | false |
+| [min_nodes](#min_nodes)                                                                     | Nodes                | No  | 0 |
+| [desired_nodes](#desired_nodes)                                                             | Nodes                | No  | 0 |
```
Comment (NikitaGandhi) [inline docs/variables.md]: value for desires_nodes been shared for both nodes and spot nodes ASG configuration. can this be separated out into two variables?

File: nodes/modules/workers/workers.tf (hunk 1 above)
Comment (tom-butler) [reply nodes/modules/workers/workers.tf]: This object accepts lists, so it should be ok

File: docs/variables.md (hunk 2 above)
Comment (tom-butler) [reply docs/variables.md]: This is an issue, but It should probably be addressed in a separate change

**Pull request merged**